pkg/agent/proxy/proxier.go

```diff
@@ -233,49 +248,82 @@ func getBindingProtoForIPProto(isIPv6 bool, protocol corev1.Protocol) binding.Pr
+			return false
 		}
 	}
 
-	// Update the reference counter of Endpoints.
+	// Update the reference counter of Endpoints and remove them from the installed Endpoints of the ServicePortName.
 	for _, endpoint := range staleEndpoints {
-		key := endpointKey(endpoint, bindingProtocol)
+		key := endpointKey(endpoint, protocol)
 		count := p.endpointReferenceCounter[key]
 		if count == 1 {
 			delete(p.endpointReferenceCounter, key)
-			klog.V(2).InfoS("Endpoint was removed", "Endpoint", endpoint.String(), "Protocol", bindingProtocol)
+			klog.V(2).InfoS("Endpoint was removed", "Endpoint", endpoint.String(), "Protocol", protocol)
 		} else {
 			p.endpointReferenceCounter[key] = count - 1
-			klog.V(2).InfoS("Stale Endpoint is still referenced by other Services, decrementing reference count by 1", "Endpoint", endpoint.String(), "Protocol", bindingProtocol)
+			klog.V(2).InfoS("Stale Endpoint is still referenced by other Services, decrementing reference count by 1", "Endpoint", endpoint.String(), "Protocol", protocol)
 		}
+		delete(p.endpointsInstalledMap[svcPortName], endpoint.String())
 	}
 
-	return nil
+	return true
+}
+
+func (p *proxier) addNewEndpoints(svcPortName k8sproxy.ServicePortName, protocol binding.Protocol, newEndpoints map[string]k8sproxy.Endpoint) bool {
+	var endpointsToAdd []k8sproxy.Endpoint
+
+	// Get all Endpoints whose reference counter is 0, and these Endpoints should be added.
+	for _, endpoint := range newEndpoints {
+		key := endpointKey(endpoint, protocol)
+		count := p.endpointReferenceCounter[key]
+		if count == 0 {
+			endpointsToAdd = append(endpointsToAdd, endpoint)
+			klog.V(2).InfoS("Endpoint will be added", "Endpoint", endpoint.String(), "Protocol", protocol)
+		}
+	}
+
+	// Add flows for these Endpoints.
+	if len(endpointsToAdd) != 0 {
+		if err := p.ofClient.InstallEndpointFlows(protocol, endpointsToAdd); err != nil {
+			klog.ErrorS(err, "Error when installing Endpoints flows for Service", "ServicePortName", svcPortName)
+			return false
+		}
+	}
+
+	// Update the reference counter of Endpoints.
+	for _, endpoint := range newEndpoints {
+		p.endpointsInstalledMap[svcPortName][endpoint.String()] = endpoint
+		key := endpointKey(endpoint, protocol)
+		p.endpointReferenceCounter[key] = p.endpointReferenceCounter[key] + 1
+	}
+	return true
 }
 
 func serviceIdentityChanged(svcInfo, pSvcInfo *types.ServiceInfo) bool {
 	return svcInfo.ClusterIP().String() != pSvcInfo.ClusterIP().String() ||
 		svcInfo.Port() != pSvcInfo.Port() ||
-		svcInfo.OFProtocol != pSvcInfo.OFProtocol ||
-		svcInfo.NodePort() != pSvcInfo.NodePort() ||
-		svcInfo.ExternalPolicyLocal() != pSvcInfo.ExternalPolicyLocal()
+		svcInfo.OFProtocol != pSvcInfo.OFProtocol
+}
+
+func serviceExternalAddressesChanged(svcInfo, pSvcInfo *types.ServiceInfo) bool {
+	return svcInfo.NodePort() != pSvcInfo.NodePort() || !reflect.DeepEqual(svcInfo.LoadBalancerIPStrings(), pSvcInfo.LoadBalancerIPStrings())
```

Comment:
nit: you can use https://pkg.go.dev/golang.org/x/exp/slices#Equal now instead of DeepEqual

Reply:
Thanks for the pointer. I saw we don't have other dependency on this package. Given it's declared experimental and unreliable, I used "k8s.io/utils/strings/slices" which has almost exactly same code.